metagpt/provider/bedrock_api.py

```diff
@@ -0,0 +1,141 @@
+        self.__client = self.__init_client("bedrock-runtime")
+        self.__provider = get_provider(self.config.model)
+        logger.warning(
+            "Amazon bedrock doesn't support asynchronous now")
+
+    def __init_client(self, service_name: Literal["bedrock-runtime", "bedrock"]):
+        """initialize boto3 client"""
+        # access key and secret key from https://us-east-1.console.aws.amazon.com/iam
+        self.__credentital_kwargs = {
+            "aws_secret_access_key": self.config.secret_key,
+            "aws_access_key_id": self.config.access_key,
+            "region_name": self.config.region_name
+        }
+        session = boto3.Session(**self.__credentital_kwargs)
+        client = session.client(service_name)
+        return client
+
+    @property
+    def client(self):
+        return self.__client
+
+    @property
+    def provider(self):
+        return self.__provider
+
+    def list_models(self):
+        """list all available text-generation models
+
+        ```shell
+        ai21.j2-ultra-v1                    Support Streaming:False
+        meta.llama3-70b-instruct-v1:0       Support Streaming:True
+        ……
+        ```
+        """
+        client = self.__init_client("bedrock")
+        # only output text-generation models
+        response = client.list_foundation_models(byOutputModality="TEXT")
+        summaries = [f'{summary["modelId"]:50} Support Streaming:{summary["responseStreamingSupported"]}'
+                     for summary in response["modelSummaries"]]
+        logger.info("\n"+"\n".join(summaries))
+
+    def invoke_model(self, request_body: str) -> dict:
+        response = self.__client.invoke_model(
+            modelId=self.config.model, body=request_body
+        )
+        usage = self._get_usage(response)
+        self._update_costs(usage)
+        response_body = self._get_response_body(response)
+        return response_body
+
+    def invoke_model_with_response_stream(self, request_body: str) -> EventStream:
+        response = self.__client.invoke_model_with_response_stream(
+            modelId=self.config.model, body=request_body)
+        usage = self._get_usage(response)
+        self._update_costs(usage)
+        return response
+
+    @property
+    def _const_kwargs(self) -> dict:
+        model_max_tokens = get_max_tokens(self.config.model)
+        if self.config.max_token > model_max_tokens:
+            max_tokens = model_max_tokens
+        else:
+            max_tokens = self.config.max_token
+
+        return {
+            self.__provider.max_tokens_field_name: max_tokens,
+            "temperature": self.config.temperature
+        }
+
+    def completion(self, messages: list[dict]) -> str:
+        request_body = self.__provider.get_request_body(
+            messages, self._const_kwargs)
+        response_body = self.invoke_model(request_body)
+        completions = self.__provider.get_choice_text(response_body)
+        return completions
```

Comment:
completion return response structure not text from get_choice_text.  see acompletion_text in BaseLLM